pkg/model/iam.go

```diff
@@ -51,134 +51,161 @@ const RolePolicyTemplate = `{
 }`
 
 func (b *IAMModelBuilder) Build(c *fi.ModelBuilderContext) error {
-	// Collect the roles in use
-	var roles []kops.InstanceGroupRole
+	// Collect managed Instance Group roles
+	managedRoles := make(map[kops.InstanceGroupRole]bool)
+
+	// Collect Instance Profile ARNs and their associated Instance Group roles
+	sharedProfileARNsToIGRole := make(map[*string]kops.InstanceGroupRole)
 	for _, ig := range b.InstanceGroups {
-		found := false
-		for _, r := range roles {
-			if r == ig.Spec.Role {
-				found = true
+		if ig.Spec.IAM != nil && ig.Spec.IAM.Profile != nil {
+			if matchingRole, ok := sharedProfileARNsToIGRole[ig.Spec.IAM.Profile]; ok {
+				if matchingRole != ig.Spec.Role {
+					return fmt.Errorf("Found IAM instance profile assigned to multiple Instance Group roles %v and %v: %v",
+						ig.Spec.Role, sharedProfileARNsToIGRole[ig.Spec.IAM.Profile], ig.Spec.IAM.Profile)
+				}
+			} else {
+				sharedProfileARNsToIGRole[ig.Spec.IAM.Profile] = ig.Spec.Role
 			}
+		} else {
+			managedRoles[ig.Spec.Role] = true
 		}
-		if !found {
-			roles = append(roles, ig.Spec.Role)
+	}
+
+	// Generate IAM tasks for each shared role
+	for profileARN, igRole := range sharedProfileARNsToIGRole {
+		iamName, err := findCustomAuthNameFromArn(profileARN)
+		if err != nil {
+			return fmt.Errorf("unable to parse instance profile name from arn %q: %v", profileARN, err)
 		}
+		err = b.buildIAMTasks(igRole, iamName, c, true)
 	}
 
-	// Generate IAM objects etc for each role
-	for _, role := range roles {
-		name := b.IAMName(role)
+	// Generate IAM tasks for each managed role
```

Comment:
Running a test on this it seems there's a logic bug somewhere. If you have more than 1 Instance Group using the same IAM role it errors out. This is important since in a HA setup each master will have it's own "Instance Group" and will share an IAM role. People may also choose to have multiple node Instance Groups sharing the same IAM role.

`task.go:57] found duplicate tasks with name "IAMRole/KubernetesNodeRole": *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"KubernetesNodeRole","RolePolicyDocument":{"Name":"","Resource":{}},"Shared":true,"ExportWithID":"nodes"} and *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"KubernetesNodeRole","RolePolicyDocument":{"Name":"","Resource":{}},"Shared":true,"ExportWithID":"nodes"}`

Reply:
Strange, it should only error if the same instance profile is used for multiple Instance Group Roles (master and node, node and bastion, etc.). Multiple instance groups using the same instance profile shouldn't fail if they are the same IGR. I'll take a closer look and let you know when I have a fix.

Reply:
Tested this and looks like it's fixed.

Reply:
> Tested this and looks like it's fixed.

This is still an issue in v1.17 and 1.18 and has not been fixed :(
Opened an issue no this #9852